src/service/resize-observer.js

```diff
@@ -0,0 +1,167 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from '../utils/array';
+import {
+  registerServiceBuilder,
+  registerServiceBuilderInEmbedWin,
+} from '../service';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ObserverCallbackDef;
+
+/**
+ * Pools resize observers. Eventually the hope is that the call sites can
+ * instantiate their own resize observers. But currently there are some
+ * performance issues and pooling is desired. For more info, see
+ * https://bugs.chromium.org/p/chromium/issues/detail?id=1157544
+ *
+ * @implements {../service.Disposable}
+ */
+export class ResizeObserverImpl {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+
+    /** @private @const {!ResizeObserver} */
+    this.observer_ = new win.ResizeObserver((e) => this.processEntries_(e));
+
+    /** @private {?WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+    this.lastSizes_ = new WeakMap();
+
+    /** @private {?WeakMap<!Element, !ObserverCallbackDef>} */
+    this.callbacks_ = new WeakMap();
+
+    /** @private @const */
+    this.microtask_ = Promise.resolve();
+  }
+
+  /** @override */
+  dispose() {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    this.callbacks_ = null;
+    this.lastSizes_ = null;
+    this.observer_.dispose();
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!ObserverCallbackDef} callback
+   */
+  observe(element, callback) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    let callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      callbacks = [];
+      this.callbacks_.set(element, callbacks);
+      this.observer_.observe(element);
+    }
+    if (!pushIfNotExist(callbacks, callback)) {
+      return;
+    }
+    const lastSize = this.lastSizes_.get(element);
+    if (lastSize) {
+      this.microtask_.then(() => callCallbackNoInline(callback, lastSize));
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!ObserverCallbackDef} callback
+   */
+  unobserve(element, callback) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    const callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.callbacks_.delete(element);
+      this.lastSizes_.delete(element);
+      this.observer_.unobserve(element);
+    }
+  }
+
+  /**
+   * @param {!Array<!ResizeObserverEntry>} entries
+   * @private
+   */
+  processEntries_(entries) {
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, clientRect} = entries[i];
+      let seenTarget = false;
```

Comment:
this is to ensure it only gives the latest size? are you confident about the ordering of the array? (newest first)

Reply:
The ordering is per spect. But the newest are last. That's why I'm going through array in the reverse order.